[WIP] Enable login timeout

AyushExel opened the pull request:
Description
-----------
This PR enables setting a timeout in the prompt choices part of `wandb.login()`

### Usage:
* `wandb.login() # defaults to 30s timeout`
* `wandb.login(timeout=10) # times out after 10 seconds of no response`

## Implementation details:
Timing out thread didn't work with inputs. This implementation uses `select` and `termios` to timeout `stdin` event in uninx/linux and `msvcrt` for windows

Testing
-------
Tox tests failed as  `wandb.utils.propt_choices` is monkey patched. I haven't messed with those files yet.

Release Notes
-------------

Below, please enter user-facing release notes as one or more bullet points.
If your change is not user-visible, write `NO RELEASE NOTES` instead, with no bullet points.

------------- BEGIN RELEASE NOTES ------------------



------------- END RELEASE NOTES --------------------

vanpelt commented on wandb/sdk/lib/input_timeout.py:
```diff
@@ -0,0 +1,72 @@
+from wandb.errors import InputTimeoutError
+
+DEFAULT_TIMEOUT = 30.0
+INTERVAL = 0.05
```
We can likely make this larger, maybe 0.1?

vanpelt reviewed (changes requested):
Let's add a test.  We'll need to unpatch `prompt_choices` and ensure the test thinks it has access to stdin then just assert DRYRUN is chosen with a short timeout.  Happy to pair if needed.

raubitsj commented:
@AyushExel Can we discuss this in #client? I dont get this change at all.

AyushExel commented:
@raubitsj Yes agreed. We'll revert this to not change the default behavior. I've started a thread in #client with more context about this request.

AyushExel commented:
@vanpelt I've addressed the review requests. There wasn't a clean way to enable stdin in pytest to I've commented out that test. I wrote another test that mocks the stdin value. The codecov is still failing though.

vanpelt commented on tests/test_stdin_timeout.py:
```diff
@@ -0,0 +1,31 @@
+    except InputTimeoutError as e:
+        assert e.message == timeout_log
+    capmanager.resume_global_capture()
+"""
```
What's up with this test?

AyushExel replied:
This is a hacky way of enabling stdin in pytest. It's not recommended to use because these low level pytest APIs can change with updates.

The bigger issue here is that the tests in this file are not being run by in tox. That's why codecov is failing. running `ox -e py37 -- tests/test_stdin_timeout.py` says that no tests were run and codecov didn't pickup any data. Do I need to put this file somewhere in the configs for tests to pick this up?
![Screenshot from 2021-08-14 12-40-11](https://user-images.githubusercontent.com/15766192/129438350-e60e39b0-900f-44ec-9e6f-8d0ca7d80170.png)

vanpelt replied:
pytest determines which functions are tests by looking for method names prepended with "test_".  Just change your cases to be named like all of the other test cases.

AyushExel replied:
@vanpelt got it thanks. Updated the test to only check timeout enabled cases. Windows test passes and gets covered in codecov. But linux/mac tests fail with a permission errors. Again, this only happens with pytest. login timeout works fine in manual tests. 
![Screenshot (78)](https://user-images.githubusercontent.com/15766192/129619100-f5f3fa44-c471-4a18-81ce-1484c3d64d9c.png)

vanpelt replied:
I imagine this is not related to pytest, instead it's related to linux or docker default permissions.  I would do some googling to see if you can find any other examples of PermissionError when registering Epoll selectors.

AyushExel replied:
After some investigation, I think it's related to pytest not allowing some os operations. suspending cap manager enables stdin but registering a selector might require some other workaround. 
I've tested the same test with pytest as the root user. It gets the same `operation not allowed` error. I've asked this on pytest forum but I'm not sure how swift their response will be.
Another way is to test this part of code independent of pytest, like a python script. Is it possible under current settings?

AyushExel replied:
@vanpelt ^^

vanpelt replied:
You should be able to use the new yea [functional testing framework ](https://github.com/wandb/client/tree/master/functional_tests) to test this.  I'm not super familiar with it, but it should be pretty straight forward after looking at the examples.  It may be tricky to convince yea that you have a tty.  The other option is to just launch a subprocess that runs code from within `/fixtures` and makes assertions based on the stdout but you won't get credit for the coverage without using the functional_tests

vanpelt commented on wandb/util.py:
```diff
@@ -973,16 +975,35 @@ def _prompt_choice():
+        return TIMEOUT_CODE
+
+
+def prompt_choices(choices, allow_manual=False, input_timeout=None):
```
To minimize the risk of the funky signal logic breaking users environments, let's just use the old logic when input_timeout is None.

AyushExel replied:
done

AyushExel commented:
@raubitsj  
**Context** - Enabling stdin with pytest and selectors wasn't working so pytest tests failed on linux/mac. So, I'm trying to cover the changes via yea test as Vanpelt suggested. [more details in the thread above]
Based on that, I have 2 doubts:
* Functional tests aren't being run in CI on windows env. I didn't get any credit in codecov for the windows stdin timeout function and there aren't any yea logs(smileys) in the windows CI test.
* Is there a way to mark specific parts to code to get codecov credits? I'm asking this because even if the windows timeout function gets codecov credit, there are a few lines of code that cannot be tested without manually passing the inputs. (We're only testing for the timeout case and not the cases where user inputs something)
* Finally, this might not be a big deal but I'm seeing something weird in the yea test that I wrote( id 0.0.6). The assertion passes on my local tests but fails in CI.
![Screenshot (80)](https://user-images.githubusercontent.com/15766192/130195136-781fb5ee-4f75-44fe-9250-ddc982ed03e5.png)
![Screenshot (81)](https://user-images.githubusercontent.com/15766192/130195150-cc2ec3bf-0445-4075-afae-96a41e1eae9e.png)

raubitsj commented on wandb/sdk/wandb_login.py:
```diff
@@ -36,7 +36,9 @@ def _handle_host_wandb_setting(host: Optional[str], cloud: bool = False) -> None
 
-def login(anonymous=None, key=None, relogin=None, host=None, force=None):
+def login(
+    anonymous=None, key=None, relogin=None, host=None, force=None, login_timeout=None
```
This should just be called "timeout" here.

To map this to login_timeout do something like:
wandb_settings.py:
```
    def _apply_login(
        self, args: Dict[str, Any], _logger: Optional[_EarlyLogger] = None
    ) -> None:
        param_map = dict(key="api_key", host="base_url", timeout="login_timeout")
```

AyushExel replied:
Thanks. This is done.

raubitsj commented on wandb/sdk/lib/stdin_timeout.py:
```diff
@@ -0,0 +1,66 @@
+import sys
```
New files in lib should get some type annotations..  Add your file to mypy.ini and annotate what makes sense, dont need everything, but it would be nice if it was tested.  For example we have one here:

```
[mypy-wandb.sdk.lib.telemetry]
disallow_untyped_defs = True
disallow_untyped_calls = True
```

AyushExel replied:
I've annotations in the stdin_timeout.py file

raubitsj commented on wandb/sdk/lib/stdin_timeout.py:
```diff
@@ -0,0 +1,66 @@
+else:
+    import time
+
+    stdin_timeout = windows_stdin_timeout
```
All this code is a bit scary... I will have to look closer.  This has the potential of breaking all wandb users if it is wrong, not just users of login(timeout=)

AyushExel replied:
Thanks. This can definitely wait if that's the case. It'd be nice to have but not at the risk of breaking existing projects.

raubitsj reviewed (changes requested):
Can you fix the merge conflicts.  will give this another look after requested changes.
